---
title: "RcwlMeta"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{RcwlMeta}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(RcwlMeta)
```

## Example
```{r}
library(RcwlPipelines)
cwlSearch(c("bwa", "align"))
bwaAlign <- RcwlPipelines::cwlLoad("pl_bwaAlign")

bwaAlign <- addMeta(cwl = bwaAlign,
                    label = "align",
                    doc = "align reads to reference genome",
                    inputLabels = c("threads", "readgroup", "reference", "read1", "read2"),
                    inputDocs = c("number of threads", "read groups", "reference genome", "read pair1", "read pair2"),
                    stepLabels = c(bwa = "bwa"),
                    stepDocs = c(bwa = "bwa alignment"))

cat(meta2md(bwaAlign))
```

## Example 2: 

```{R}
rcp <- ReUseData::recipeLoad("gencode_annotation")
meta(rcp)
rcp1 <- addMeta(cwl = rcp,
                label = "",
                doc = "An empty description line", 
                inputLabels = c("input label1", "input label2"),
                inputDocs = c("input description 1", "input description 2"), 
                outputLabels = c("output label1"),
                outputDocs = c("output description 1"), 
                extensions = list(author = "recipe author's name",
                                  url = "http://ftp.ebi.ac.uk/pub/databases/gencode/",
                                  date = as.character(Sys.Date()),
                                  example = "An example chunk of code that was coverted into md format"))
meta(rcp1)
cat(meta2md(rcp1))
```
